docs/_design/map-controller.md

```diff
@@ -0,0 +1,404 @@
+may not be available.
+Without that input object, the hook author presumably wouldn't have enough
+information to decide on an updated desired state anyway.
+We can reexamine this if users come up with compelling use cases.
+
+[tombstone hook]: #tombstone-hook
+
+### Status Aggregation
+
+One notable omission from the map hook, as compared with the sync hook from
+CompositeController, is that the user does not return any status object.
+That's because each map hook invocation only sends enough context to process a
+single input object and its associated output objects.
+The hook author therefore doesn't have enough information to compute the overall
+status of the parent object.
+
+We could define another hook to which we send all inputs and outputs for a given
+parent, and ask the user to return the overall status.
+However, that would defeat one of the main goals of MapController because such a
+monolithic hook request could get quite large for the type of use cases we
+expect for a controller that says, "do this for every X," and also because that
+would place the burden of aggregating status across the whole collection onto
+the user.
+
+Instead, Metacontroller will compute an aggregated status for the collection
+based on some generic rules:
+
+For each input resource, we will report the number of matching objects we
+observed as a status field on the parent object, named after the plural
+resource name.
+
+The exact format will be an implementation detail, but for example it might
+look like:
+
+```yaml
+status:
+  inputs:
+    persistentvolumeclaims:
+      total: 20
+  ...
+```
+
+For each output resource, we will report the total number of objects owned by
+this parent across all map keys.
+In addition, we will automatically aggregate conditions found on output objects,
+and report how many objects we own with that condition set to `True`.
+
+For example:
+
+```yaml
+status:
+  ...
+  outputs:
+    volumesnapshots:
+      total: 100
+      ready: 97
+  ...
+```
+
+## Hooks
+
+### Map Hook
+
+We call the map hook to translate an input object into zero or more output
+objects.
+
+#### Map Hook Request
+
+| Field | Description |
+| ----- | ----------- |
+| `controller` | The whole MapController object, like what you might get from `kubectl get mapcontroller <name> -o json`. |
+| `parent` | The parent object, like what you might get from `kubectl get <parent-resource> <parent-name> -o json`. |
+| `mapKey` | An opaque string that uniquely identifies the group of outputs that belong to this input object. |
+| `input` | The input object, like what you might get from `kubectl get <input-resource> <input-name> -o json`. |
+| `outputs` | An associative array of output objects that the parent already created for the given input object. |
+
+#### Map Hook Response
+
+| Field | Description |
+| ----- | ----------- |
+| `outputs` | A list of JSON objects representing all the desired outputs for the given input object. |
+
+### Tombstone Hook
```

Comment:
I think tombstone hook is a more flexible design (e.g., user wants to retain type A output object, but not other types of output objects), but wish to verify whether its role can be technically achieved by existing propagation policy and finalizer feature of garbage collection.

Reply:
Do you mean, could we use finalizers as an alternative to the tombstone hook? If so, I think the problem is that the objects would go into a "pending deletion" state, which cannot be undone if the controller later decides to keep the object (e.g. a matching input PVC gets recreated, so the SnapshotSchedule controller decides to keep the "detached" VolumeSnapshot output objects after all).

Reply:
Noted. Tombstone hook is necessary because "pending deletion" state cannot be undone.